include/deal.II/numerics/matrix_creator.templates.h

```diff
@@ -1332,94 +1335,97 @@ namespace MatrixCreator
-                                        (coefficient_vector_values[point](component_i) * u * v * weight);
-                                    }
-                                copy_data.cell_vector.back()(i) += rhs_values_system[point](component_mapping[component_i])
-                                                                   * v
-                                                                   * weight;
-                              }
-                          }
-                      }
+                    const FiniteElement<dim,spacedim> &base = fe.base_element(fe.component_to_base_index(comp).first);
+                    const unsigned int bcomp = fe.component_to_base_index(comp).second;
+
+                    if (!base.conforms(FiniteElementData<dim>::H1) &&
+                        base.conforms(FiniteElementData<dim>::Hdiv))
+                      for (unsigned int point=0; point<fe_values.n_quadrature_points; ++point)
+                        normal_adjustment[point][comp] = fe_values.normal_vector(point)[bcomp]
+                                                         * fe_values.normal_vector(point)[bcomp];
                   }
-                else  //      if (coefficient == 0)
-                  for (unsigned int point=0; point<fe_values.n_quadrature_points; ++point)
-                    {
-                      const double weight = fe_values.JxW(point);
-                      for (unsigned int i=0; i<fe_values.dofs_per_cell; ++i)
+
+                for (unsigned int point=0; point<fe_values.n_quadrature_points; ++point)
+                  {
+                    const double weight = fe_values.JxW(point);
+                    for (unsigned int i=0; i<fe_values.dofs_per_cell; ++i)
+                      if (fe_is_primitive)
                         {
-                          const double v = fe_values.shape_value(i,point);
                           for (unsigned int j=0; j<fe_values.dofs_per_cell; ++j)
-                            if (fe.system_to_component_index(i).first ==
-                                fe.system_to_component_index(j).first)
-                              {
-                                const double u = fe_values.shape_value(j,point);
-                                copy_data.cell_matrix.back()(i,j) += (u * v * weight);
-                              }
+                            {
+                              if (fe.system_to_component_index(i).first
+                                  == fe.system_to_component_index(j).first)
+                                {
+                                  copy_data.cell_matrix.back()(i,j)
+                                  += coefficient_vector_values[point](fe.system_to_component_index(i).first)
+                                     * fe_values.shape_value(i,point)
+                                     * fe_values.shape_value(j,point)
+                                     * weight;
```

Comment:
wouldn't you want to do the normal adjustment here as well? I know that in most cases the element will not be primitive *and* only Hdiv/Hcurl conforming, but it's conceivable that someone comes up with such an element.

Reply:
Sorry, I don't quite get it. Do you mean that one can come up with a primitve element that is Hdiv/Hcurl (which I doubt is possible) or non-primitive element that is H1 (which indeed might be)? 
And same question as above, as it is not done in the non-hp `create_boundary_mass_matrix_1` version would you want to change that one as well?

Reply:
Yes, there are plenty of non-primitive elements that are H1 conforming. But I really intended the comment to be about the possibility of primitive Hdiv elements. I don't know if they exist, or if they are possible, but I typically just like erring on the side of being safe for future extensions. Here, that would mean either putting the adjustment in, or having an assertion that verifies that the adjustment is not necessary.